### ISSUE TITLE
DocFX build fails occasionally with  System.NullReferenceException

### ISSUE THREAD

**Issue by Raj270892** [dotnet/docfx#3445]
**.NET Command Line Tools Version Used**: 2.1.200

**.NET Core Shared Framework Version Used**: 2.0.7

**mono Version Used**:  5.12.0.226

**Linux Version Used**:  Ubuntu 16.04.5

**Kernel Version Used**:  4.15.0-34-generic

**gitlab-runner Version Used**:  11.2.0

**Description**:
I have dcFX project integrated with visual studio which builds successfully on windows but fails occasionally on linux server. I am using gitlab runner ci to build my docfx project every half an hour and I can see it passes sometimes while fails most of the time. It throws following error message which does not give much information to debug 

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
    at System.Runtime.Remoting.Lifetime.ClientSponsor.Close () [0x00023] in <71d8ad678db34313b7f718a414dfcb25>:0 
    at System.Runtime.Remoting.Lifetime.ClientSponsor.Finalize () [0x00000] in <71d8ad678db34313b7f718a414dfcb25>:0
/home/gitlab-runner/builds/c02422e5/0/Common/SoftwareDocumentation/packages/Symbotic.Documentation.Build.1.4.0/build/Symbotic.Documentation.Build.targets(97,5): error MSB3073: The command ""/home/gitlab-runner/builds/c02422e5/0/Common/SoftwareDocumentation/packages/Symbotic.Documentation.Build.1.4.0/build/../tools/docfx.exe" build "/home/gitlab-runner/builds/c02422e5/0/Common/SoftwareDocumentation/Source/Symbotic.Documentation/docfx.json" --log "/home/gitlab-runner/builds/c02422e5/0/Common/SoftwareDocumentation/Source/Symbotic.Documentation/buildlog-docs.log" --logLevel "Verbose"" exited with code 255. [/home/gitlab-runner/builds/c02422e5/0/Common/SoftwareDocumentation/Source/Symbotic.Documentation/Symbotic.Documentation.csproj]

```

**Expected Behavior**: Build Should pass all the time

**Actual Behavior**: Build pass sometime and fails occasionally.

**Comment by Raj270892** [comment]
build log is huge and I thought this is the only peace of information helpful out of that

I can collect all information by any other mean if you want just let me know how should I do it

**Comment by Raj270892** [comment]
I was wondering to know if there is any update on this issue.

**Comment by superyyrrzz** [comment]
Seems there is not enough information to figure out the problem. Can you provide all the logs since docfx.exe starts to run?

**Comment by Raj270892** [comment]
I have updated my previous post and also here in I have attached the log file 

[raw.txt](https://github.com/dotnet/docfx/files/2501921/raw.txt)

Let me know if you need anything else

**Comment by Raj270892** [comment]
Is there any update on this issue?

**Comment by KalleOlaviNiemitalo** [comment]
I think it is a bug in the class library of Mono. The finalizer of ClientSponsor does not call Close in .NET Framework, and it should not do so in Mono either. Calling Close from there is not useful because, if the finalizer is being called, then .NET Remoting no longer has a reference to the ClientSponsor; if the lease tries to call ISponsor.Renewal regardless, then it will get an exception and unregister the sponsor.

In addition, ClientSponsor.InitializeLifetimeService just returns null in .NET Framework, but it appears to call MarshalByRefObject.InitializeLifetimeService in Mono. I think this means the Mono ClientSponsor instance will become inaccessible via .NET Remoting when its initial lease expires, even if it is still referenced within its AppDomain.

Until those Mono bugs are fixed, you could try copying the ClientSponsor class from Reference Source to DocFX, removing the BCLDebug.Trace calls, and using that instead of the Mono implementation. That should work more reliably.

(The Mono implementation of ClientSponsor is better than the .NET Framework implementation in one respect: ClientSponsor.Unregister actually removes the object from the Hashtable. In .NET Framework, the objects accumulate until Close is called. That eventually leads to difficult-to-understand remoting exceptions when hash collisions occur and Hashtable tries to call the Equals method of a remote object whose lease already expired. OTOH, a Dictionary<MarshalByObject, ILease> with a ReferenceEquality comparer would entirely avoid remoting those Equals calls.)

**Comment by herohua** [comment]
@superyyrrzz , is it possible for users to workaround such errors by disabling app domain and run everything in the default domain?

**Comment by yufeih** [comment]
This issue should be addressed in the [v2.60.0](https://github.com/dotnet/docfx/discussions/8334) release that runs on .NET core.
